### DISCUSSION TITLE
Important notice: set your audio buffers to 64x3!!

### DISCUSSION THREAD

**Discussion by rerdavies** [rerdavies/pipedal#164] [General]
Important notice: The 1.2.34 release chanced the default value for audio buffers to 64x2. This is seems to be an exceptionally bad choice, that causes drastic audio underruns when cpu use is high, and sometimes even stops the realtime audio thread altogether.

For USB audio adapters,  buffers should always be set to x3. This is an old piece of lore that used to apply to Jack Audio devices, but I thought did not apply to ALSA devices. It turns out that it very much does!

If your audio buffers are set to 64x2, please change them to either 64x3, or 32x3. I think you should not experience audio dropouts using 32x3 audio buffers; but 64x3 seems to be bullet-proof even under very high CPU load.

**Comment by rerdavies** [comment]
I've been playing, a bit, with using 16x4 buffer settings. It seems surprisingly stable, and might have lower round-trip latency than 32x3.(If anyone should know for sure, it should be me, but I don't). Superficially, it seems pretty stable even with 50%+ CPU loads.

I've been tied down with development tasks, so I haven't had a chance to try this in actual live performance to see if it's rock solid. I do development work on my actual Pi, and PiPedal requires large buffers when the pi is under heavy load. I have, however, let PiPedal run overnight with no underruns at all over 8 hour periods. So I think it might actually be a good setting.

Let me know what your experiences are.

**Comment by davidofchatham** [comment]
Is this still your current advice? In the documentation I've seen 32x4 recommended, but I came across the discussion about NAM threading where 64x was recommended, and now this.

My old E-MU 0404 USB adapter does seem to "favor" the x3 settings slightly when using the latency test, but thus far I've had it set at 32x4, the lowest setting that seemed stable. 

I've noticed that sometimes the latency test will repeatedly run successfully for a particular configuration, but switching PiPedal to that configuration immediately results in overruns, which is a little odd but may be expected if you know what's going on.

**Comment by rerdavies** [comment]
I currently think 32x4 is the best starting point.  Slightly more resistant
to underruns, slightly more CPU overhead , for most of my test devices.
Although that may depend on what hardware you are using, which plugins you
use, and how heavily loaded your CPU is.

Latency tests run with no CPU load, so it isn't a great surprise to me that
configurations that that run fine under the latency tests are not
practically usable once plugins are loaded.

**Reply by davidofchatham** [reply]
Okay, thank you!

Does threading still require a 64x configuration to work efficiently? I'm running PiPedal on a six-core AMD FX processor, so it seems like it may be advantageous to enable threading if I use multiple NAM models at once.

**Comment by rerdavies** [comment]
No, it does not require 64x buffers anymore.

**Reply by davidofchatham** [reply]
Great — thanks!

**Reply by Bl4ck-Eyes** [reply]
So... What would be suggested on Rpi5 8gb RAM?
32x2? 32x3? Something else?

What would be the best way to determine it? 
Thanks in advance.
Fra.
